### PR TITLE
Add git-cop gem 

### PR DESCRIPTION
**Pull request by Ana06**
Adding `git-cop` gem to improve our commits.

The enabled rules are:

- `commit_author_email`: Ensures author email address exists.

- `commit_body_bullet`: enforce `-` for lists.

- `commit_body_leading_space`: Ensures there is blank line between the commit subject and body.

- `commit_body_line_length`: Ensures each line of the commit body is no longer than 72 characters.

- `commit_body_phrase`: Avoid meaningless words/phrases: *obviously*, *basically*, *simply*, *of course*, *everyone knows* and *easy*.

- `commit_subject_length`: Ensures the commit subject length is no more than 50 characters.

- `commit_subject_prefix`: Ensure that the commit has one or more tags (*api*, *app*, *backend*, *ci*, *dist*, *doc*, *webui*) alphabetically ordered and that the commit subject start by one of:
  - Fix
  - Add
  - Change
  - Update
  - Remove
  - Refactor
  - Merge
  - Split
  - Enable
  - Disable

  Take into account that for being able to use `commit_subject_prefix` with this specification I needed to add all the possible combinations in the configuration file. It will be possible to use regular expressions in git-cop in the future, but currently there is not other way. I used the following code for generating that:

  ```
  tags = ["[api]",
          "[backend]",
          "[ci]",
          "[dist]",
          "[doc]",
          "[webui]"]
  prefixes = ["Fix",
              "Add",
              "Change",
              "Update",
              "Remove",
              "Refactor",
              "Merge",
              "Split",
              "Enable",
              "Disable"]

  tags_list = (0..tags.length).map{
                |index| tags.combination(index).map(&:join)
              }.flatten.reject!(&:empty?)
  result = "- " + prefixes.map{
             |prefix| (tags_list.map{
               |tags| "\"" + tags + " " + prefix + "\""
             }).join("\n- ")
           }.join("\n- ")
  puts result
  ```

git-cop would be run with this rule in Travis together with our other linters cops.

I also add a *How to write proper commit messages* section to the our contributing file which include the git-cop current configuration. I also documented `dist` tag, as we are already using it but it is not in our contributing file.

@openSUSE/open-build-service all of us should agree on this, so please take a look :wink:

### CODE REVIEW

**Comment by Ana06** [comment]
I still have to fix something regarding the configuration file.

**Comment by ChrisBr** [comment]
Why did you breakt the test suite in the first place? I would prefer to first get it fixed before we use a monkey patched gem to be honest ...

**Comment by Ana06** [comment]
@ChrisBr 

> Why did you breakt the test suite in the first place? I would prefer to first get it fixed before we use a monkey patched gem to be honest ...

It was an accident.... I will fix their test suite, send a PR to the gem and then we can use the new release of the gem. We are not going to use a monkey patched gem, I just wanted to let you know that it can not be merged yet.

But probably the most difficult part to address in this PR is to agree in the rules we want to check, that's why now that I am sure we can use the gem, I wanted to open the discussion even if I have some things to finish.

File: src/api/Gemfile [hunk 1]
```diff
@@ -119,6 +119,8 @@ group :test do
   gem 'rails-controller-testing'
   # To generate random data
   gem 'rantly', '>= 1.1.0'
+  # To ensure consistent Git commits
+  gem 'git-cop', git: 'https://github.com/Ana06/git-cop.git', branch: 'travis-ci'
```
Comment (hennevogel) [inline src/api/Gemfile]: It's not possible to use gems from git for us, we need to be able to package them.

**Comment by marcus-h** [comment]
Some thoughts below...

On 2017-07-05 05:56:00 -0700, Ana María Martínez Gómez wrote:
> Adding `git-cop` gem to improve our commits.
> 
> The enabled rules are:
> 
> - `commit_author_email`: Ensures author email address exists.
> 
> - `commit_body_bullet`: enforce `-` for lists.
> 
> - `commit_body_leading_space`: Ensures there is blank line between the commit subject and body.
> 
Just to be sure: this does not mean that the body has to start with a
leading (white)space, right?

> - `commit_body_line_length`: Ensures each line of the commit body is no longer than 72 characters.
> 
Makes sense, but there is at least one exception: exceptions/backtraces:
IMHO, it would be "ridiculous" to wrap them after 72 chars.

> - `commit_body_phrase`: Avoid meaningless words/phrases**: *obviously*, *basically*, *simply*, *of course*, *everyone knows* and *easy*.
> 
Hmm I wouldn't call them meaningless per se... In fact, I would
presume that the author used them intentionally (e.g., to stress
or to put the emphasis on something...)

> - `commit_subject_length`: Ensures the commit subject length is no more than 50 characters.
> 
Makes sense, but sometimes hard to achieve. IMHO, it is more important to
have a meaningful subject, than a subject that does not tell you anything,
but does not exceed the length limit...
Maybe check for <= 50 + delta, where delta > 0 is some (large)
threshold?

> - `commit_subject_prefix`: Ensure that the commit has one or more tags (*api*, *app*, *backend*, *ci*, *dist*, *doc*, *webui*) alphabetically ordered and that the commit subject start by one of:
>   - Fix
>   - Add
>   - Change
>   - Update
>   - Remove
>   - Refactor
>   - Merge
>   - Split
>   - Enable
>   - Disable

The tag check makes a lot of sense, but checking the first word of the
subject is, IMHO, nonsense.

I wouldn't enfore all these rules by technical means, but rely on common
sense during the review process:)
Nevertheless, it would be good to have them in the CONTRIBUTING.md.

File: src/api/Gemfile (hunk 1 above)
Comment (Ana06) [inline src/api/Gemfile]: I know, I am going to send a PR to the gem with the changes and then we will use the next release of the gem. But I created the PR to discuss what cops we want to use.

**Comment by Ana06** [comment]
@hennevogel 

> `+ commit_subject_prefix (but please just the tags, I don't care too much how you phrase this)`

Why just the tags? The other is about to write the commit message in imperative mode, which is even mention in [Github good commit messages guide](https://github.com/erlang/otp/wiki/Writing-good-commit-messages#do)

> `+ commit_author_email (Is that even possible with github?)`

I guess it is not, but I guess it doesn't hurts to check it.

```
> - commit_body_bullet (I don't care)
> - commit_body_leading_space (I don't care)
> - commit_body_line_length (I don't care)
> - commit_body_phrase (I don't care)
> - commit_subject_length (I don't care)
```
I don't care means that you are fine if we enable then or that you don't want to have them?

**Comment by hennevogel** [comment]
>> `+ commit_subject_prefix (but please just the tags, I don't care too much how you phrase this)`

> Why just the tags? The other is about to write the commit message in imperative mode, which is even mention in [Github good commit messages guide](https://github.com/erlang/otp/wiki/Writing-good-commit-messages#do)

Because guiding someone is something else than enforcing rules by technical means. We can all be role models and follow this guide, I'm all for it! But making contributions 'fail' because you did not use a specific *voice* in your commit message is too much out of balance for me.

>> [...more rules...]

> I don't care means that you are fine if we enable then or that you don't want to have them?

I don't care means I don't want those rules enforced either, pretty much out of the same reason.

**Comment by Ana06** [comment]
@marcus-h 


> > - `commit_body_leading_space`: Ensures there is blank line between the commit subject and body.
> 
> Just to be sure: this does not mean that the body has to start with a
> leading (white)space, right?

The name of the cop is not very appropriate, would be change in the gem. It means that you have to leave a blank line as separation between the commit title and the body. This is actually really important, from [Github good commit messages guide](https://github.com/erlang/otp/wiki/Writing-good-commit-messages):

> The blankline separating the summary from the body is critical (unless you omit the body entirely); tools like rebase can get confused if you run the two together.

> > - `commit_body_line_length`: Ensures each line of the commit body is no longer than 72 characters.
> > 
> Makes sense, but there is at least one exception: exceptions/backtraces:
> IMHO, it would be "ridiculous" to wrap them after 72 chars.

There a severity feature currently being implemented in the gem. If we change the severity of this cop to warning it would tell you that you should wrap it in 72, but won't fail. But to be honest I would prefer to enforce the 72 character with an error, as otherwise `git log` won't be render properly in an 80-column terminal.

> > - `commit_body_phrase`: Avoid meaningless words/phrases**: *obviously*, *basically*, *simply*, *of course*, *everyone knows* and *easy*.
> > 
> Hmm I wouldn't call them meaningless per se... In fact, I would
> presume that the author used them intentionally (e.g., to stress
> or to put the emphasis on something...)

I would suggest disabling this until the warning option is implemented, and then we could use the warning option, so it doesn't fails. Because I agree that this can be more flexible.

> > - `commit_subject_length`: Ensures the commit subject length is no more than 50 characters.
> > 
> Makes sense, but sometimes hard to achieve. IMHO, it is more important to
> have a meaningful subject, than a subject that does not tell you anything,
> but does not exceed the length limit...
> Maybe check for <= 50 + delta, where delta > 0 is some (large)
> threshold?

But if the subject is longer than 50 characters in most cases you won't be able to read it complete, what is even worse than that it miss information. You can always add more information in the description. And this is not something I just like, from git commit [manpage:](https://www.kernel.org/pub/software/scm/git/docs/git-commit.html#_discussion)

> it’s a good idea to begin the commit message with a single short (less than 50 character) line summarizing the change, followed by a blank line and then a more thorough description. The text up to the first blank line in a commit message is treated as the commit title, and that title is used throughout Git.

> > - `commit_subject_prefix`: Ensure that the commit has one or more tags (*api*, *app*, *backend*, *ci*, *dist*, *doc*, *webui*) alphabetically ordered and that the commit subject start by one of:
> >   - Fix
> >   - Add
> >   - Change
> >   - Update
> >   - Remove
> >   - Refactor
> >   - Merge
> >   - Split
> >   - Enable
> >   - Disable
> 
> The tag check makes a lot of sense, but checking the first word of the
> subject is, IMHO, nonsense.

ok, @hennevogel also things the same, so only tags. :+1

**Comment by Ana06** [comment]
@hennevogel 
> But making contributions 'fail' because you did not use a specific voice in your commit message is too much out of balance for me.

I really think it is much easy to contribute in a project with proper commit message, even if it implies some more work. I really don't think that this will discourage people of contributing. It is the same with style checkers, they also makes contributions 'fail', but it is always much comfortable to contribute to a proper written code. 

Anyway @marcus-h also doesn't like the commit subject starting by  a fixing word, so let put this out.

> I don't care means I don't want those rules enforced either, pretty much out of the same reason.

I would really like to enable `commit_body_leading_space`, because I already told to @marcus-h it is really important. from Github good commit messages guide:

> The blankline separating the summary from the body is critical (unless you omit the body entirely); tools like rebase can get confused if you run the two together.

I also would like to enable `commit_subject_length`, as I told to @marcus-h if the subject is longer than 50 characters in most cases you won't be able to read it complete, what is even worse than that it miss information. You can always add more information in the description. And this is not something I just like, from git commit manpage:

> it’s a good idea to begin the commit message with a single short (less than 50 character) line summarizing the change, followed by a blank line and then a more thorough description. The text up to the first blank line in a commit message is treated as the commit title, and that title is used throughout Git.

And with the rest rules, what about enabling then with warning severity (when this is implemented, it is currently being developed in the gem), so they tell you that it is wrong but doesn't fail?

**Comment by evanrolfe** [comment]
Is there a way we can make this advisory rather than mandatory? Then maybe after a couple weeks of using this we can take a vote from the team to see if we want to make this mandetory or not?

**Comment by Ana06** [comment]
@evanrolfe 

> Is there a way we can make this advisory rather than mandatory? Then maybe after a couple weeks of using this we can take a vote from the team to see if we want to make this mandetory or not?

Currently no, I suggested to the maintainer of the gem to introduce a warning feature and he is currently implementing it. I guess it would be available in the following days, because I saw the branch and it is almost finished, but I am not sure.

**Comment by Ana06** [comment]
[add git-cop gem](https://trello.com/c/Bv3EaCNH/448-add-git-cop-gem)

**Comment by hennevogel** [comment]
> commit_body_leading_space

This rule would be okay for me, the name confused me too, sorry

> commit_body_line_length or won't be render properly in an 80-column terminal.

I'm sorry but this what I meant with off balance. Because of some arbitrary rule that doesn't apply to any of us (all your terminals are re-sizable) you want to trigger work and frustration. For the submitter to change the message, for for the CI circle etc. etc.

Look just because something makes sense you don't have to do it. You have to weigh the consequences of it against what you gain.

> And with the rest rules, what about enabling then with warning severity (when this is implemented, it is currently being developed in the gem), so it tells you that it is wrong but doesn't fail?

That would be awesome. However I would argue that the test suite is the wrong time/place then, but better than nothing.

**Comment by Ana06** [comment]
Okey, then what about enabling the following:

- `commit_author_email`: Ensures author email address exists. **error**

- `commit_body_bullet`: enforce `-` for lists. **warning**

- `commit_body_leading_space`: Ensures there is a blank line between the commit subject and body. **error**

- `commit_body_line_length`: Ensures each line of the commit body is no longer than 72 characters. **warning**

- `commit_subject_length`: Ensures the commit subject length is no more than 50 characters. **warning**

- `commit_subject_prefix`: Ensure that the commit has one or more tags (*api*, *app*, *backend*, *ci*, *dist*, *doc*, *webui*) alphabetically ordered **error**

@openSUSE/open-build-service please take a look!

**Comment by DavidKang** [comment]
@Ana06, I agree :+1:

**Comment by eduardoj** [comment]
@Ana06 , I also agree with you proposed changes:

- `commit_author_email`: Ensures author email address exists. **error**
- `commit_body_bullet`: enforce - for lists. **warning**
-  `commit_body_leading_space`: Ensures there is a blank line between the commit subject and body. **error**
-  `commit_body_line_length`: Ensures each line of the commit body is no longer than 72 characters. **warning**
- `commit_body_phrase`: Avoid meaningless words/phrases: obviously, basically, simply, of course, everyone knows and easy. **warning**
-  `commit_subject_length`: Ensures the commit subject length is no more than 50 characters. **warning**
-  `commit_subject_prefix`: Ensure that the commit has one or more tags (api, app, backend, ci, dist, doc, webui) alphabetically ordered. **error**

**error**: will make the test fail.
**warning**: will show a warning, and let the test continue.

Nice!

**Comment by bgeuken** [comment]
> marcus-h commented 17 hours ago
>
> Makes sense, but there is at least one exception: exceptions/backtraces:
IMHO, it would be "ridiculous" to wrap them after 72 chars.

I agree. I wouldn't want to reformat backtraces that I add to a commit message.

**Comment by bgeuken** [comment]
> DavidKang commented 39 minutes ago
>
> @Ana06, I agree

Me too. Let's try it out.

**Comment by bgeuken** [comment]
One thing regarding the subject length... keep in mind that this is already shortened because of the tags.

**Comment by Ana06** [comment]
@bgeuken 
> One thing regarding the subject length... keep in mind that this is already shortened because of the tags.

Yes, actually a better approach would be not to include the tags in the title, but at the end of the body.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -25,12 +25,62 @@ In particular, this community seeks the following types of contributions:
...
 * [backend] - Changes in the perl-written backend of OBS
 * [ci]      - Changes that affect our test suite
+* [dist]    - Modifies something inside /dist directory
 * [doc]     - Any documentation related changes
+* [webui]   - Changes in webui related parts of app/model/ and lib/ as well as app/controllers/webui/ and it's views
+
+In case of having more than one tag, they should be **alphabetically** ordered.
+
+
+### Useful commit subject
```
Comment (hennevogel) [inline CONTRIBUTING.md]: *nitpick* How about we just point to the google guidelines?

Comment (Ana06) [inline CONTRIBUTING.md]: what guidelines?

Comment (hennevogel) [inline CONTRIBUTING.md]: Sorry not google, github -> https://github.com/erlang/otp/wiki/Writing-good-commit-messages

File: src/api/Gemfile.lock [hunk 3]
```diff
@@ -1,3 +1,14 @@
+GIT
```
Comment (hennevogel) [inline src/api/Gemfile.lock]: We still can't do that. Are you going to resolve this or should we just close the PR until there is a new release?

Comment (Ana06) [inline src/api/Gemfile.lock]: There is already a new release, but the travis support is not working properly. Travis behave different in a PR than in a branch. In the branch works, so I need the PR to find out was going on.

Comment (hennevogel) [inline src/api/Gemfile.lock]: Any news?

**Comment by Ana06** [comment]
I'd already tested what I needed and it is working, but I would open a new PR because I find the conversation difficult to follow in this PR and because many things changed in the gem, so the code looks already quite different.

**Pull request closed without merging**